compiler/src/iree/compiler/API2/Internal/Diagnostics.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/API2/Internal/Diagnostics.h"
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/IR/BuiltinAttributes.h"
+
+// Much of the formatting logic below was adapted from the
+// SourceMgrDiagnosticHandler but without the logic for printing actual source
+// contents and other adaptations to detach it from a SourceMgr.
```

Comment:
(and River and I was talking about how it doesn't actually do what we want and clang has a much more complete one ... so definitely an interesting area :))